All markdown files are properly formatted now

nickytonline opened the pull request:
This PR implements #162, and adds a precommit hook via [lint-staged](https://github.com/okonet/lint-staged/) and [husky](https://github.com/typicode/husky). For the time being, I have only added an entry to lint-staged to format markdown (md) files. As well, all the md files have been formatted using the prettier settings that I copied from Storybook's settings.

Built everything locally and it looks good. I just did a smoke test navigating to some pages.

![learnstorybook.com screeshot](https://user-images.githubusercontent.com/833231/66687352-5fc81e00-ec50-11e9-8058-a0828819b2b3.png)

Also, Skeletor approves of this PR.

![Skeletor approves](https://media.giphy.com/media/aWRWTF27ilPzy/giphy.gif)

nickytonline commented on .prettierignore:
```diff
@@ -0,0 +1,2 @@
+yarn.lock
```
We never want prettier to format these files.

nickytonline commented on .prettierrc:
```diff
@@ -0,0 +1,7 @@
+{
```
Copied from https://github.com/storybookjs/storybook/blob/next/.prettierrc

kylesuss replied:
Is it possible to make this formatting work with the prettier config that we have already here?

https://github.com/chromaui/chromatic/blob/next/.eslintrc.js#L31

The eslint config is already setup with these params.

nickytonline replied:
Oops, didn’t see that. I will update it later tonight.

nickytonline commented on package.json:
```diff
@@ -40,6 +40,17 @@
     "storybook": "start-storybook -s ./static -p 6006",
     "test": "echo \"Error: no test specified\" && exit 1"
   },
+  "husky": {
```
The precommit magic.

nickytonline commented on package.json:
```diff
@@ -40,6 +40,17 @@
+    }
+  },
+  "lint-staged": {
+    "*.md": [
```
Other file types can be added here in another PR.

nickytonline commented on .eslintrc.js:
```diff
@@ -23,17 +25,7 @@ module.exports = {
-        jsxBracketSameLine: false,
-      },
-    ],
+    'prettier/prettier': ['warn', prettierConfig],
```
This is imported now, so that files that are non-JS can use the same prettier config.

nickytonline replied:
I am curious as to why eslint is configured to only warn about prettier. I imagine once a config for JS files is added to the lint-staged configuration, this would no longer be required?

kylesuss replied:
I think the initial logic was that we didn't want to discourage people from contributing because of it. But yeah, given the changes you have implemented here which help reinforce the standards, I think we can change it eventually.

nickytonline commented on .prettierrc.js:
```diff
@@ -0,0 +1,8 @@
+module.exports = {
```
This is the config that get's imported into .eslintrc.js

nickytonline commented:
Alright, I updated things to use the existing prettier config. Let me know if that's all good. That was the first time I've seen a prettier config in an eslint config which is why I didn't think to look there. ;)

nickytonline commented:
Yeah, sharing seemed to make the most sense to me for the prettier confit.

I’ll rerun the batch prettier formatting for the non-prettier formatted *.md files. I think I used the wrong glob when I ran it, so it didn’t format all the files. I’ll update the PR this afternoon.

And as you mentioned, the only way for markdown files to get formatted is on the precommit hook. If you have an editor like VS Code you can have them formatted on save with the prettier extension, but there is no guarantee that everyone contributing would be using VS Code, so the precommit hook is a fail safe.

As mentioned previously, once this PR is merged, we can use the precommit hook for other file types as well: *.jsx, *.js, *.json etc.

nickytonline commented:
So the reason that particular file would not prettify is because of the location of the array square brackets for contributors. I lined them up with those of the `authors` section and then prettier ran as usual. That was not obvious. It looks like that is the only file that wasn't prettified. I'll push that change momentarily.

nickytonline commented:
Hmm even with that fix there is still parts that did not convert to single quotes. I'll continue to investigate.

nickytonline commented:
Ahh, I see what it is. It's mentioned in a prettier issue. See https://github.com/prettier/prettier/tree/21733e441dc01f7d85b483edee92b7e7507bfd9a#quotes. It's by design.

> If the number of quotes outweighs the other quote, the quote which is less used will be used to format the string - Example: "I'm double quoted" results in "I'm double quoted" and "This \\"example\\" is single quoted" results in 'This "example" is single quoted'.

nickytonline commented:
I went through the other markdown files that have double quotes, and they are either respecting the rule I mentioned previously or they are markup and stay as double quotes instead of single quotes.

nickytonline commented on content/design-systems-for-developers/index.md:
```diff
@@ -1,67 +1,68 @@
+title: 'Design Systems for Developers'
+description: 'Discover how to build and maintain design systems using Storybook.'
+heroDescription: 'A guide that teaches professional developers how to transform component libraries into design systems and set up the production infrastructure used by leading frontend teams.'
 overview: "Design systems power the frontend teams of Shopify, IBM, Salesforce, Airbnb, Twitter,  and many more. This guide for professional developers examines how the smartest teams engineer design systems at scale and why they use the tools they use. We'll walk through setting up core services, libraries, and workflows to develop a design system from scratch."
```
As mentioned in the main convo, this is how prettier works. Since there is a single quote in. the text, it will keep the double quotes for the string.

kylesuss commented on .prettierignore:
```diff
@@ -0,0 +1,2 @@
+yarn.lock
+package-lock.json
```
Nice. Haha we aren't supposed to have both of these lockfiles, so I will submit a small PR to remove the `package-lock.json`.

kylesuss commented on content/intro-to-storybook/react/zh-TW/composite-component.md:
```diff
@@ -1,19 +1,19 @@
 ---
 
-上一章我們構建了第一個元件; 本章 我們學習 擴充套件構建TaskList的任務列表. 讓我們將 元件組合 在一起,看看在引入更多複雜性時會發生什麼.
+上一章我們構建了第一個元件; 本章 我們學習 擴充套件構建 TaskList 的任務列表. 讓我們將 元件組合 在一起,看看在引入更多複雜性時會發生什麼.
```
@danielduan or @chinanf-boy -- pinging you here since you contributed the original Chineses translations. Prettier is formatting some of these files slightly differently. I recognize there are 2 Chinese translations and it appears to happen in both of them:

Seems like the common change is to insert a space around English words and Chinese characters. So:

`擴充套件構建TaskList的任務列表`

becomes

`擴充套件構建 TaskList 的任務列表`

The same happens in the `zh-CN` translation. Does this "break" the translation at all? Or is it an acceptable change? If it breaks things, I think we can probably ignore these files in Prettier.

Thanks 🙇

nickytonline commented:
Sounds good. I’ll get to this later this evening.

nickytonline commented:
Alright, I've rebased and fixed the merge conflicts @kylesuss. So this should be good to go. As you mentioned, just waiting to hear about the Chinese translations.

domyen commented:
@kylesuss @nickytonline I suggest we move forward with auto-formatting Chinese. According to this [article](https://www.hillslearning.com/2011/06/15/writing-in-chinese-spacing/) it's fine to add a space between English words and Chinese characters.

kylesuss commented:
@domyen nice find. Interestingly, that article seems to imply that the old way was wrong! Ha. Yes let's move forward on this and circle back if need be.
